exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/NumericEquiDepthHistogram.java

```diff
@@ -76,129 +73,188 @@ public NumericEquiDepthHistogram(int numBuckets) {
-                // value is exactly equal to the first bucket's start point so we take the ratio 1/bucket_width
-                long totalFilterRows = (long) (1 / (buckets[first + 1] - buckets[first]) * numRowsPerBucket);
-                double selectivity = (double) totalFilterRows / totalRows;
-                return selectivity;
-              } else {
-                // predicate is 'column < value' and value is equal to first bucket's start point, so none of
-                // the rows qualify
-                return SMALL_SELECTIVITY;
-              }
-            }
-
-            // *** End of boundary conditions ****
-
-            int n = getContainingBucket(value, numBuckets);
-            if (n >= 0) {
-              // all buckets to the left will be fully covered
-              int coveredBuckets = n;
-              long coveredRows = numRowsPerBucket * coveredBuckets;
-              // num matching rows in the current bucket is a function of (value - start_point_of_bucket)
-              long partialRows = (long) ((value - buckets[n]) / (buckets[n + 1] - buckets[n]) * numRowsPerBucket);
-              long totalFilterRows = partialRows + coveredRows;
-              double selectivity = (double)totalFilterRows / totalRows;
-              return selectivity;
-            } else {
-              // value does not exist in any of the buckets
-              return SMALL_SELECTIVITY;
-            }
+        Double value = getLiteralValue(filter);
+        if (value == null) {
+          unkownFilterList.add(filter);
+        } else {
+          // get the operator
+          SqlOperator op = ((RexCall) filter).getOperator();
+          Range range;
+          switch (op.getKind()) {
+            case GREATER_THAN:
+              range = Range.greaterThan(value);
+              currentRange = currentRange.intersection(range);
+              break;
+            case GREATER_THAN_OR_EQUAL:
+              range = Range.atLeast(value);
+              currentRange = currentRange.intersection(range);
+              break;
+            case LESS_THAN:
+              range = Range.lessThan(value);
+              currentRange = currentRange.intersection(range);
+              break;
+            case LESS_THAN_OR_EQUAL:
+              range = Range.atMost(value);
+              currentRange = currentRange.intersection(range);
+              break;
+            default:
+              break;
           }
         }
       }
     }
-    return null;
+    return currentRange;
+  }
+
+  private long getSelectedRows(final Range range) {
+    final int numBuckets = buckets.length - 1;
+    double startBucketFraction = 1.0;
+    double endBucketFraction = 1.0;
+    long numRows = 0;
+    int result;
+    Double lowValue = null;
+    Double highValue = null;
+    final int first = 0;
+    final int last = buckets.length - 1;
+    int startBucket = first;
+    int endBucket = last;
+
+    if (range.hasLowerBound()) {
```

Comment:
There is code duplication for lowValue and highValue except the inverted comparisons. Maybe we can refactor the common portion out?

Reply:
There are some key differences in the comparisons and the the bucket ranges that make it difficult to consolidate and even if I was to do it it would look somewhat contrived.  Let me know if you have something specific in mind.  For now I have left it as is.